### PR TITLE
Bluetooth: host: Fix GATT clear bond information bugs

### LINKED ISSUE #23961: CCC does not get cleared when CONFIG_BT_KEYS_OVERWRITE_OLDEST is enabled
**Describe the bug**
When CONFIG_BT_KEYS_OVERWRITE_OLDEST is enabled, a warning gets thrown by bt_gatt saying no space to store CCC cfg.

**To Reproduce**
1. `git apply prj.conf.patch && west build -b nrf52_pca10040 samples/bluetooth/peripheral_hr`
2. `west flash`
3. bond with the peripheral using nrfconnect and subscribe to the CCC notification
4. disconnect and perform step 3. with a different device (bluetoothctl)
5. disconnect and perform step 3 again with the first device. The CCC cannot be written to the device and the error message appears

**Expected behavior**
Expected is that the CCC can be normally subscribed to without warning messages from `bt_gatt`

**Impact**
This leads to all notifications being broken once in this situation. If the device is power cycled the state is returned to normal.

**Screenshots or console output**
```
[00:00:55.689,941] <inf> hrs: HRS notifications enabled
Disconnected (reason 0x13)
[00:01:51.021,789] <inf> hrs: HRS notifications disabled
Connected
[00:02:06.110,473] <inf> bas: BAS Notifications enabled
[00:03:09.010,406] <inf> hrs: HRS notifications enabled
[00:03:38.810,302] <wrn> bt_ctlr_hci: type: 0x12
Disconnected (reason 0x13)
[00:03:51.810,699] <inf> bas: BAS Notifications disabled
[00:03:51.810,760] <inf> hrs: HRS notifications disabled
Connected
[00:03:58.308,197] <wrn> bt_ctlr_hci: type: 0x14
[00:04:11.032,073] <wrn> bt_gatt: No space to store CCC cfg
```

**Environment (please complete the following information):**
 - OS: manjaro
 - Toolchain gcc-arm-non-eabi-9-2019-q4-major
 - Commit SHA or Version used: `commit 9518bd19b7a736bd649fa2fc8d8d52967c0b486f (HEAD, tag: zephyr-v2.2.0, tag: v2.2.0)`

**Additional context**
Short discussion with Joakim revelead this could be an issue in CONFIG_BT_KEYS_OVERWRITE_OLDEST, as it's quite a new feature

patch:
```
diff --git a/samples/bluetooth/peripheral_hr/prj.conf b/samples/bluetooth/peripheral_hr/prj.conf
index 3dfa8796b8..f65f1a0913 100644
--- a/samples/bluetooth/peripheral_hr/prj.conf
+++ b/samples/bluetooth/peripheral_hr/prj.conf
@@ -8,3 +8,7 @@ CONFIG_BT_GATT_BAS=y
 CONFIG_BT_GATT_HRS=y
 CONFIG_BT_DEVICE_NAME="Zephyr Heartrate Sensor"
 CONFIG_BT_DEVICE_APPEARANCE=833
+
+CONFIG_BT_KEYS_OVERWRITE_OLDEST=y
+CONFIG_BT_WHITELIST=y
+CONFIG_BT_MAX_PAIRED=1
\ No newline at end of file
```

### PR DESCRIPTION
**Pull request by joerchan**
Fix CCC cfg not cleared when overwriting oldest bond. Calling
bt_unpair with a pointer to the key will result in the key addr being
memset to zero and bt_gatt_clear is called with an zero-set address.
This happens because unpair (hci_core.c) calls bt_keys_clear before
calling bt_gatt_clear.

Fixes: #23961

Fix Service Changed configuration stored in flash not deleted in flash
when calling bt_gatt_clear from bt_unpair.
When clearing the check for "is bonded" should not be made. If the
bond information is already removed this check will fail.
When clearing the check for "modified" should not be made, clearing
is in itself a modification, and should always be made.

### CODE REVIEW

**Comment by joerchan** [comment]
CC: @mfiumara

**Comment by joerchan** [comment]
@jhedberg This could have been fixed by re-ordering bt_gatt_clear and bt_keys_clear. I'm wondering if we should do both fixes. Assuming re-ordering has no negative effects.

**Comment by jhedberg** [comment]
@joerchan `bt_gatt_clear()` ends up calling `bt_addr_le_is_bonded()` when clearing Service Changed information, so I think the order must be kept (would be good to have a code comment though so no-one goes and changes this)

**Review by jhedberg — DISMISSED** [review]

**Comment by Vudentz** [comment]
> @joerchan `bt_gatt_clear()` ends up calling `bt_addr_le_is_bonded()` when clearing Service Changed information, so I think the order must be kept (would be good to have a code comment though so no-one goes and changes this)

That is just removing the entry from the settings, not sure if that is already done implicitly by bt_unpair, it doesn't looks like it does anything related to settings though, it seems it just clear the keys entry so actually doing the bt_gatt_clear before bt_keys_clear should make sure the settings storage is cleanup properly.

**Comment by joerchan** [comment]
@Vudentz It looks like the SC would not be cleared from flash then. The modified check would also affect this.

**Comment by jhedberg** [comment]
@Vudentz @joerchan so what's the consensus on this PR? Do we merge it it it needs additional changes?

**Comment by joerchan** [comment]
> @Vudentz @joerchan so what's the consensus on this PR? Do we merge it it it needs additional changes?

I would say merge this one. And I'll look if there is a separate problem and submit a PR if that is the case.

File: subsys/bluetooth/host/gatt.c [hunk 1]
```diff
@@ -252,44 +252,68 @@ static void sc_store(struct gatt_sc_cfg *cfg)
...
-			if (err) {
-				BT_ERR("failed to delete SC (err %d)", err);
-			} else {
-				BT_DBG("deleted SC for %s (%s)",
-				       bt_addr_le_str(&cfg->peer),
-				       log_strdup(key));
-			}
+static void sc_clear(struct bt_conn *conn)
+{
+	if (!bt_addr_le_is_bonded(conn->id, &conn->le.dst)) {
```
Comment (jhedberg) [inline subsys/bluetooth/host/gatt.c:302]: Could you flip the main- and else-branches so that you don't have to do the negation for the condition?

**Review by jhedberg — APPROVED** [review]

**Review by Vudentz — APPROVED** [review]
LGTM

**Pull request merged**